### DISCUSSION TITLE
yolov5/tutorials/model_ensembling/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#13169] [Docs]
# yolov5/tutorials/model_ensembling/

Learn how to ensemble YOLOv5 models for improved mAP and Recall! Clone the repo, install requirements, and start testing and inference.

https://docs.ultralytics.com/yolov5/tutorials/model_ensembling/

**Comment by TanWeeKang** [comment]
Hi, is there a way to ensemble models using python instead of CLI? I am trying to test if the ensembled models would perform better or the current ones of individual will do.

**Reply by pderrenger** [reply]
Hi there!

Yes, you can ensemble models using Python instead of the CLI. Here's a quick example to get you started:

```python
from ultralytics import YOLO

# Load your models
model1 = YOLO('yolov8n.pt')
model2 = YOLO('yolov8m.pt')

# Perform inference with both models
results1 = model1('path/to/image.jpg')
results2 = model2('path/to/image.jpg')

# Combine results (example for bounding boxes)
combined_boxes = results1.boxes + results2.boxes

# Process combined results as needed
# ...

# Optionally, visualize combined results
# combined_results.plot()
```

This approach allows you to load multiple models, run inference, and then combine their results programmatically. For more detailed guidance, you can refer to the [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling/) section in the YOLOv5 documentation.

Feel free to experiment and see if the ensembled models improve your performance! 🚀

**Reply by TanWeeKang** [reply]
Hi, when i run it says list has no attribute boxes on the combined_boxes line

**Reply by pderrenger** [reply]
Hello!

It seems like you're encountering an issue because the `results1` and `results2` objects don't directly expose a `boxes` attribute that can be combined using a simple addition operation. Instead, you should access the bounding boxes through the `boxes` attribute of the `results` object, which provides a structured way to handle them.

Here’s how you can modify your code to combine the bounding boxes from two models correctly:

```python
from ultralytics import YOLO

# Load your models
model1 = YOLO('yolov8n.pt')
model2 = YOLO('yolov8m.pt')

# Perform inference with both models
results1 = model1('path/to/image.jpg')
results2 = model2('path/to/image.jpg')

# Access and combine bounding boxes
combined_boxes = results1.boxes.xyxy + results2.boxes.xyxy

# Now you can process combined_boxes as needed
```

Make sure that the operations you perform on `combined_boxes` are appropriate for the data structure, which is typically a tensor. If you need further manipulation (like concatenation or stacking), you might need to use functions from libraries like `torch`.

If you have any more questions or need further assistance, feel free to ask!

**Comment by PhucPhamSy** [comment]
Dear author, is there a way to ensemble 2 models yolo9e-seg and yolo9c-seg, then visualize only ensemble's result?

**Reply by glenn-jocher** [reply]
@PhucPhamSy yes, you can ensemble the YOLO9e-seg and YOLO9c-seg models and visualize the ensemble's results. To do this, you can append both models to the `--weights` argument in your inference command. This will aggregate the predictions from both models into a single output. For more detailed instructions, please refer to the [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling/) guide. If you encounter any issues, ensure you're using the latest versions of the models and the Ultralytics package.

**Reply by PhucPhamSy** [reply]
Thank you! How about the way for doing with Python script instead of CLI?

**Reply by glenn-jocher** [reply]
To ensemble YOLO9e-seg and YOLO9c-seg models using a Python script, you can load both models and aggregate their predictions. Here's a concise example:

```python
from ultralytics import YOLO

# Load the models
model1 = YOLO("yolo9e-seg.pt")
model2 = YOLO("yolo9c-seg.pt")

# Run inference on an image
results1 = model1("path/to/image.jpg")
results2 = model2("path/to/image.jpg")

# Combine results (e.g., averaging bounding boxes, masks)
ensemble_results = (results1 + results2) / 2

# Visualize the ensemble results
ensemble_results.show()
```

This script loads both models, runs inference, combines the results, and visualizes the ensemble output. For more details, refer to the [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling/) guide.

**Reply by PhucPhamSy** [reply]
I got the error like this, it seems like can't combination because not same data type.
`ensemble_results = (result1 + result2) / 2`
TypeError: unsupported operand type(s) for /: 'list' and 'int'

**Reply by glenn-jocher** [reply]
@PhucPhamSy the error you're encountering is due to the fact that `results1` and `results2` are lists, and you cannot directly perform arithmetic operations on lists. To ensemble the models correctly, you need to aggregate their predictions in a way that combines the results appropriately. You can achieve this by processing the results from each model and then combining them. Ensure you are using the latest versions of the models and the Ultralytics package to avoid any compatibility issues. For detailed guidance on model ensembling, please refer to the [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling/) guide.

**Reply by PhucPhamSy** [reply]
thank you! I try to use detect.py in the Model Ensembling guide. And got this error like this
`Traceback (most recent call last):
  File "/raid/data/phuc/segmentation/SDI-yolo/yolov5/detect.py", line 425, in <module>
    main(opt)
  File "/raid/data/phuc/segmentation/SDI-yolo/yolov5/detect.py", line 420, in main
    run(**vars(opt))
  File "/home/ubuntu/anaconda3/envs/yolo5/lib/python3.9/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/raid/data/phuc/segmentation/SDI-yolo/yolov5/detect.py", line 250, in run
    s += f"{n} {names[int(c)]}{'s' * (n > 1)}, "  # add to string
KeyError: 230`

**Reply by glenn-jocher** [reply]
It looks like you're encountering a KeyError related to the class index during the ensembling process. This issue often arises when the class indices in the models do not match. Please ensure that both models are trained on the same dataset with identical class indices. Additionally, verify that you are using the latest versions of the Ultralytics package and the models. If the issue persists, try running the script with a single model to isolate the problem. For further details, you can refer to the [Model Ensembling guide](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling/).

**Comment by SarvananFreeStone** [comment]
Hi, in python how do we ensemble a yolo object detection with a following yolo image classification, any module created with optimized/parallel computing

**Reply by glenn-jocher** [reply]
@SarvananFreeStone to ensemble a YOLO object detection model with a YOLO image classification model, you can run them sequentially in your Python script. First, perform object detection to identify regions of interest, then pass these regions to the classification model. While there's no specific module for optimized parallel computing in this context, you can use Python's multiprocessing or threading for efficiency. For more on model ensembling, you might find the YOLOv5 documentation helpful.

**Comment by jackyy** [comment]
Hi, when the models are ensembled are the weights of the models averaged, or the predictions from each model are averaged?

**Reply by glenn-jocher** [reply]
@jackyy the predictions from each model are averaged during ensembling, not the weights. This aggregation of predictions helps improve the overall performance. You can find more details on model ensembling in the [Ultralytics YOLOv5 documentation](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling/).
